### DISCUSSION TITLE
Can we write android related business logic

### DISCUSSION THREAD

**Discussion by raaviswami** [cashapp/zipline#1337] [Q&A]
For example getting list of installed applications  using packageManager , 

  ```
 val applications = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            context.packageManager.getInstalledApplications(
                PackageManager.ApplicationInfoFlags.of(
                    PackageManager.GET_META_DATA.toLong()
                )
            )
        } else {
            @Suppress("DEPRECATION") context.packageManager.getInstalledApplications(PackageManager.GET_META_DATA)
        }
```

**Comment by JakeWharton** [comment]
No. You would have to put that code inside a service and bind it from the host to be made available to the guest code inside the VM.

**Reply by raaviswami** [reply]
Wondering if you have any  such example that uses Android SDK or Android components
Use cases can be 
1) Get the list of installed applications on device 
2) Get the location lat/long

**Reply by raaviswami** [reply]
are you saying we have keep android related business logic inside RealWorldClockHost(as per your clock example ) ? and access it from jsMain(Js.kt) ?

**Reply by JakeWharton** [reply]
That's right. You have to explicitly expose any APIs from the host platform into the guest (JS) code.

For example, in another repo where we use Zipline we define this interface: https://github.com/cashapp/redwood/blob/a3924943493049370b08b655bebdeb489e17db9f/samples/emoji-search/presenter-treehouse/src/commonMain/kotlin/com/example/redwood/emojisearch/treehouse/HostApi.kt

Which is then implemented on Android: https://github.com/cashapp/redwood/blob/a3924943493049370b08b655bebdeb489e17db9f/samples/emoji-search/android-views/src/main/kotlin/com/example/redwood/emojisearch/android/views/RealHostApi.kt

And iOS: https://github.com/cashapp/redwood/blob/a3924943493049370b08b655bebdeb489e17db9f/samples/emoji-search/ios-uikit/EmojiSearchApp/IosHostApi.swift
